src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/Validation/IShouldValidate.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Validation
+{
+    /// <summary>
+    /// Contract for attributes that determine whether associated properties should be validated. When the attribute is
+    /// applied to a property, the validation system calls <see cref="ShouldValidateEntry"/> to determine whether to
+    /// validate that property. When applied to a type, the validation system calls <see cref="ShouldValidateEntry"/>
+    /// for each property within that type to determine whether to validate it.
+    /// </summary>
+    public interface IShouldValidate
+    {
+        /// <summary>
+        /// Gets an indication whether the <paramref name="entry"/> should be validated.
+        /// </summary>
+        /// <param name="entry"><see cref="ValidationEntry"/> to check.</param>
+        /// <param name="parentEntry"><see cref="ValidationEntry"/> containing <paramref name="entry"/>.</param>
+        /// <returns><c>true</c> if <paramref name="entry"/> should be validated; <c>false</c> otherwise.</returns>
+        bool ShouldValidateEntry(ValidationEntry entry, ValidationEntry parentEntry);
```

Comment:
One thing that I don't love about this is that it doesn't provide you the opportunity to 'close over' the property you are attached to (if you are attached to a property) it's a parameter.

I also don't really love that we're making `ValidationEntry` more prominent.

For instance
```
[MyShouldValidate]
public class Widget
{
    public string Name { get; set; }
}

public class MyShouldValidateAttribute : Attribute
{
    public bool ShouldValidateEntry(ValidationEntry entry, ValidationEntry parentEntry)
    {
        // right here I need to look at entry.Metadata to figure out what property we're looking at
    }
}
```
But then again if you're attached to a property you will always be called with the same metadata each time. These feel like different tasks and a lot of the information that's relevant for one is irrelevant for the other.


What if we did something like this? `GetPropertyValidationFilter` would be called once for each property and could close over any metadata details that are significant.

```
public class OptionalOfTValidateAttribute : Attribute, IPropertyValidationFilter
{
    public Func<object, bool> GetPropertyValidationFilter(ModelMetadata property)
    {
        return container => property.ValueGetter(container) != default(Optional<T>);
    }
}

public class ValidateNeverAttribute : Attribute, IPropertyValidationFilter
{
    public Func<object, bool> GetPropertyValidationFilter(ModelMetadata property)
    {
        return container => false;
    }
}
```